Ability to split reconciler duties

arthurdm opened the pull request:
Fixes #63 

Adding a section that splits reconciler duties.

arthurdm commented:
hey @nebhale - when we reach agreement on this PR, could you please help with the CRD code generation update for the new flag?

scothis commented on README.md:
```diff
@@ -175,7 +177,7 @@ A Service Binding Resource **MAY** define a `.spec.mappings` which is an array o
 A Service Binding Resource **MAY** define a `.spec.env` which is an array of `EnvVar`.  An `EnvVar` object **MUST** define `name` and `key` entries.  The `key` of an `EnvVar` **MUST** refer to a binding `Secret` key name including any key defined by a `Mapping`.  The value of this `Secret` entry **MUST** be configured as an environment variable on the resource represented by `application`.
 
-A Service Binding resource **MUST** define a `.status.conditions` which is an array of `Condition` objects.  A `Condition` object **MUST** define `type`, `status`, and `lastTransitionTime` entries.  At least one condition containing a `type` of `Ready` **MUST** be defined.  The `status` of the `Ready` condition **MUST** have a value of `True`, `False`, or `Unknown`.  The `lastTransitionTime` **MU ...
+A Service Binding resource **MUST** define a `.status.conditions` which is an array of `Condition` objects.  A `Condition` object **MUST** define `type`, `status`, and `lastTransitionTime` entries.  At least one condition containing a `type` of `Ready` **MUST** be defined.  The `status` of the `Ready` condition **MUST** have a value of `True`, `False`, or `Unknown`.  The `lastTransitionTime` **MU ...
```
the last sentence was dropped. Bad merge?

> A Service Binding resource **MAY** reflect the secret projected into the application as `.status.binding.name`.

scothis replied:
oh it moved, nm

scothis commented on README.md:
```diff
@@ -364,6 +366,40 @@ If a `.spec.type` is set, the `type` entry in the binding `Secret` **MUST** be s
+
+There are scenarios where the reconciler that processes the bindings of a provisioned service is different than the reconciler that will project those bindings into the Application. In such cases the `ServiceBinding` CR author **MUST** set `.spec.application.autoProjection` to `false`, signalling the separation of reconciliation duties between `.spec.service` and `.spec.application`.  The default ...
+
+When `.spec.application.autoProjection` is set to `false`, the reconciler that processes `.spec.service` **MUST** reflect the generated secret, which is ready for Application projection, as `.status.binding.name`.  The reconciler that processes `.spec.application` can therefore watch for this particular `.status` state, and then proceed with Application projection.  When projection is completed t ...
```
Two reconcilers both handing the same resource is generally an anti-pattern that can led to races and thrashing the API Server with mutations.

![giphy](https://user-images.githubusercontent.com/302992/87332694-f68c7880-c509-11ea-9acc-2234e0697241.gif)

There is also state that can only be defined by the former reconciler. For example, if a mapping template is invalid, or if there was any other error generating the secret, that error needs to be reflected.

arthurdm replied:
Ya - I thought about introducing a new Condition of type `SecretReady`, that the former reconciler would use to indicate failure and that the latter reconciler could use to abort waiting for `.status.binding`.  Would that address your concern?

arthurdm replied:
it's `duck type` season 😆

scothis replied:
> I thought about introducing a new Condition of type SecretReady, that the former reconciler would use to indicate failure and that the latter reconciler could use to abort waiting for .status.binding. Would that address your concern?

Partially.

Adding a new condition would give a way to report incremental status for aspects of the reconciliation. This wouldn't solve the issue where there are two reconcilers actively trying to write to the same resource. While possible, this behavior is tricky to get right and requires cooperation on both ends. Given that one of those ends is open to extension, I'm not optimistic about the ecosystem being well behaved.

While less informative to watchers of the `ServiceBinding` resource, I'd rather see a single reconciler be responsible for writing the status. When that reconciler is not responsible for projecting the binding into the application, the `ServiceBinding`'s status likewise is not responsible for reporting the status of work it's not responsible for.

arthurdm replied:
> the `ServiceBinding`'s status likewise is not responsible for reporting the status of work it's not responsible for

I would like to break the assumption that a `ServiceBinding` resource is directly tied to a single processing unit.  I have always viewed `ServiceBinding` and its status as the encapsulation of a set of binding activities:
* processing the binding secret of the provisioned service
* creating mappings
* projecting the final bindings (mount and env vars)

The `ServiceBinding` resource should be agnostic to the fact that we may have 1, 2 or 3 different processors doing the work - and given that this work is sequential (never in parallel), I don't think there would be a race condition for carrying out the work and updating the different Condition types.  

Essentially, by defining the status Condition for each stage we then have well defined hand-off points of control.

scothis replied:
> I would like to break the assumption that a ServiceBinding resource is directly tied to a single processing unit. I have always viewed ServiceBinding and its status as the encapsulation of a set of binding activities

My issue is that this model goes against the current best practice to have a single reconciler for a resource. It's certainly possible to have multiple reconcilers for a single resource, but things can quickly go sideways unless every component trying to manage that resource has a clear understanding of its boundaries. But we're talking about this as an extension point, so I don't trust those external reconcilers to be well behaved. When they inevitably do misbehave, users will blame the `ServiceBinding` since that's what they created and all they can see.

An alternative approach that would be robust is to create an internal SPI resource that is created by the ServiceBinding reconciler representing the extensible sub-task. A third-party operator can manage the reconciliation of that resource, and the ServiceBinding can watch for status changes on the internal resource and report that status back out in a meaningful way. There is precedent for this model in Knative with its Ingress (ingresses.networking.internal.knative.dev) resource.  The KIngress is created by the Knative Serving controller, but the reconciler for KIngress is provided by an independent networking plugin.

arthurdm replied:
that's a good suggestion @scothis 

I wonder though:  if we have a corresponding status Condition for each of the delineated stages of `ServiceBinding`, and this is defined at the spec level, isn't that enough to provide the clear understanding of the boundaries?    

I would think that setting a particular status Condition, where you "own" that status Condition on the shared resource, would be easier to implement than creating a new internal resource + manage its lifecycle + setting its status Condition.

scothis replied:
> I wonder though: if we have a corresponding status Condition for each of the delineated stages of `ServiceBinding`, and this is defined at the spec level, isn't that enough to provide the clear understanding of the boundaries?

I don't think this behavior is good idea and don't want any suggestions or nudges towards that behavior in the core spec because it's too easy for one poor third-party implementation to cause a whole cluster to go sideways. Individual implementations are of course free to do what they see fit for their own needs by going beyond the spec.

> I would think that setting a particular status Condition, where you "own" that status Condition on the shared resource, would be easier to implement than creating a new internal resource + manage its lifecycle + setting its status Condition.

This is a case where the implementation that seems easier leads to instability. It'll ultimately be easier for everyone to do the right thing from the beginning.

scothis commented on README.md:
```diff
@@ -364,6 +366,40 @@ If a `.spec.type` is set, the `type` entry in the binding `Secret` **MUST** be s
 
+## Reconciliation Duties
+
+There are scenarios where the reconciler that processes the bindings of a provisioned service is different than the reconciler that will project those bindings into the Application. In such cases the `ServiceBinding` CR author **MUST** set `.spec.application.autoProjection` to `false`, signalling the separation of reconciliation duties between `.spec.service` and `.spec.application`.  The default ...
```
As we discussed on the issue, booleans in APIs are often problematic. With Go, defaulting the value to `true` will require the type be `*bool` which will make it harder to use and programmatically create.

arthurdm replied:
I guess I am looking purely from a spec (language agnostic) perspective, where a boolean is a simple primitive that requires less learning.  The fact that we came up with so many variations for the Enums, none unanimously accepted, hinted to me that perhaps this required a simpler solution.

scothis replied:
As you mentioned, naming is indeed hard, but it's worth taking a couple iterations to get right. What I like about an enum is that it's a natural point for extension in the future.

For example, today we support projecting into PodSpec-able resources. If down the road there is a new resource shape we want to support, or a PodTemplateSpec subresource hosted by the k8s API Server, then it would be nice to have a way to instruct to the operator which strategy to use for the projection.

arthurdm replied:
that's a fair point.  I will revisit the Enum idea.

arthurdm commented:
@scothis / @nebhale - I have now simplified the PR text and switched to using Enums.   I tried to make it clear that processing of `ServiceBinding` is shared, as 2 elements are processed by each framework.

Avni-Sharma commented on README.md:
```diff
@@ -175,7 +178,7 @@ A Service Binding Resource **MAY** define a `.spec.mappings` which is an array o
 A Service Binding Resource **MAY** define a `.spec.env` which is an array of `EnvVar`.  An `EnvVar` object **MUST** define `name` and `key` entries.  The `key` of an `EnvVar` **MUST** refer to a binding `Secret` key name including any key defined by a `Mapping`.  The value of this `Secret` entry **MUST** be configured as an environment variable on the resource represented by `application`.
 
-A Service Binding resource **MUST** define a `.status.conditions` which is an array of `Condition` objects.  A `Condition` object **MUST** define `type`, `status`, and `lastTransitionTime` entries.  At least one condition containing a `type` of `Ready` **MUST** be defined.  The `status` of the `Ready` condition **MUST** have a value of `True`, `False`, or `Unknown`.  The `lastTransitionTime` **MU ...
+A Service Binding resource **MUST** define a `.status.conditions` which is an array of `Condition` objects.  A `Condition` object **MUST** define `type`, `status`, and `lastTransitionTime` entries.  At least one condition containing a `type` of `Ready` **MUST** be defined.  The `status` of the `Ready` condition **MUST** have a value of `True`, `False`, or `Unknown`.  The `lastTransitionTime` **MU ...
```
Hi, Should we address https://github.com/k8s-service-bindings/spec/issues/75 as part of this PR?

arthurdm replied:
hey @Avni-Sharma - maybe we could have a separate PR that expanded the set of examples?

arthurdm commented:
latest commits push this work into an extension, as we define a new annotation and CRD (internal) for the reconciler separation.

scothis commented on README.md:
```diff
@@ -362,12 +367,109 @@ The `$SERVICE_BINDING_ROOT` environment variable **MUST NOT** be reset if it is
 
 Extensions are optional additions to the core specification as defined above.  Implementation and support of these specifications are not required in order for a platform to be considered compliant.  However, if the features addressed by these specifications are supported a platform **MUST** be in compliance with the specification that governs that feature.
 
+## Custom Projection
```
Is there something that tells the "A" reconciler to stop the default projection when the custom projection path is triggered?

arthurdm replied:
good point, I will add that.

arthurdm replied:
added a new commit to address that.

arthurdm commented on README.md:
```diff
@@ -362,12 +367,109 @@ The `$SERVICE_BINDING_ROOT` environment variable **MUST NOT** be reset if it is
+metadata:
+  name: account-service
+  annotations:
+    projection.service.binding: "Custom"
```
I had thought about that, but it would collide with the convention used in the [annotation extension](https://github.com/k8s-service-bindings/spec#annotation-examples), which serves a different purpose.

arthurdm replied:
I like that.  will update!

arthurdm commented on README.md:
```diff
@@ -352,7 +357,7 @@ status:
 ## Reconciler Implementation
 
-A Reconciler implementation for the `ServiceBinding` type is responsible for binding the Provisioned Service binding `Secret` into an Application.  The `Secret` referred to by `.status.binding.name` on the resource represented by `service` **MUST** be mounted as a volume on the resource represented by `application`.  If the `application` resource is managed by another Reconciler, a `ServiceBindin ...
+A Reconciler implementation for the `ServiceBinding` type is responsible for binding the Provisioned Service binding `Secret` into an Application.  The `Secret` referred to by `.status.binding.name` on the resource represented by `service` **MUST** be mounted as a volume on the resource represented by `application`.  If the `application` resource is managed by another Reconciler it is **RECOMMEND ...
```
not sure I agree with that statement.   I actually think it's important to make clear whenever an extension is available to expand on the current core context.   Extensions are optional functionality, but they are still in the spec.

nebhale replied:
You end up with logical circular references if you start doing this.  The core spec *must* be able to stand on its own, without any knowledge of an extension.  If we violate this, it would require implementations that are opting _not_ to implement an extension to have knowledge and work around that extension.

arthurdm replied:
I think the key is that this text is _recommending_ the extension for a particular scenario, not _mandating_ it.  So implementers of the core spec can happily ignore that recommendation as if it wasn't there.  But, they could also like the recommendation and look further into the extension for a more compatible implementation.

scothis replied:
Was it intentional to remove this line?

> `ServiceBinding` Implementations **SHOULD** ensure that the `Secret` volume mount configuration remains after the other Reconciler completes.

The way the new sentence is spliced is a bit off:

> If the `application` resource is managed by another Reconciler it is **RECOMMENDED** that both Reconcilers comply with the specification extension defined in [Custom Projection](#custom-projection).

The first half is talking about how the application resource's desired state is defined, the second half (custom projection) is about how the application resource's desired state is reconciled.

I would still prefer that we avoid links to extensions in the spec's core. If we want to hint to the user that there is more on the subject in an extension we can say something like:

> Specification extensions may define more specific projection behavior.

arthurdm replied:
> Was it intentional to remove this line?

Yeah, it hints at an anti-pattern.  If implementations want to do this that's fine, but the spec shouldn't advocate for it. :) 

Thanks for the feedback on the extensions - I will rework that portion in a new commit.

arthurdm replied:
In the new commit I just remove that sentence - this way the paragraph becomes more concise and stays within the core behaviour.

arthurdm commented on README.md:
```diff
@@ -362,12 +367,109 @@ The `$SERVICE_BINDING_ROOT` environment variable **MUST NOT** be reset if it is
 
 If a `.spec.type` is set, the `type` entry in the binding `Secret` **MUST** be set to its value overriding any existing value.  If a `.spec.provider` is set, the `provider` entry in the binding `Secret` **MUST** be set to its value overriding any existing value.
 
+### Binding Status
```
I set this anchor here so I can reference it from the new extension, instead of duplicating text.

scothis replied:
How about renaming it to something like `Ready Condition Status`, since this section is focused specifically on the ready condition's status and not the whole binding status.

arthurdm replied:
good idea.  updated.

arthurdm commented on README.md:
```diff
@@ -362,12 +367,109 @@ The `$SERVICE_BINDING_ROOT` environment variable **MUST NOT** be reset if it is
 
+## Custom Projection
+
+There are scenarios where the Reconciler that processes the bindings of a Provisioned Service (hereinafter referred to as `Reconciler A`) is different than the Reconciler that will project those bindings into the Application (hereinafter referred to as `Reconciler B`). To coordinate the projection between Reconcilers A and B the `ServiceBinding` CR author **MUST** set the `projection.service.bind ...
```
See previous comment about clashing with existing annotation extension.

arthurdm commented on README.md:
```diff
@@ -362,12 +367,109 @@ The `$SERVICE_BINDING_ROOT` environment variable **MUST NOT** be reset if it is
+
+There are scenarios where the Reconciler that processes the bindings of a Provisioned Service (hereinafter referred to as `Reconciler A`) is different than the Reconciler that will project those bindings into the Application (hereinafter referred to as `Reconciler B`). To coordinate the projection between Reconcilers A and B the `ServiceBinding` CR author **MUST** set the `projection.service.bind ...
+
+Reconciler A reacts to this annotation by creating a new `ServiceBindingProjection` CR which includes the necessary information for Reconciler B, who is watching for and takes ownership of new `ServiceBindingProjection` resources (see the [Custom Projection Definition](#custom-projection-definition) section below), to carry out the projection.  Reconciler A is responsible for updating `ServiceBin ...
```
my thought is that the different reconcilers would simply key off the apiVersion / Kind, rather than the annotation.  For example, the Runtime Component Operator would process any `ServiceBindingProjection` that had `application.kind` set to `RuntimeComponent`.  This would avoid the need for `ServiceBinding` authors to know all of the enums that could go into the annotation and when to use each one.

arthurdm replied:
Good discussion - it's important to explore the different use cases.  I can definitely see the value of that extra metadata for duck type. But I still go back to balancing user experience of `ServiceBinding` authors vs extra work for Reconcilers - always putting more weight to the former.

As a `ServiceBinding` author is probably easier for me to just add `projection.service.binding/type: Custom` (thanks for the suggestion!), than it is for me to know what metadata will trigger the correct Reconciler.  

From the Reconciler's (B) perspective, it should be able to check for duck type patterns just like others (like `status.binding.name`).  And if Reconciler B really wants to use metadata, maybe it relies on the authors to set the app via a labelSelector, and check for `ServiceBindingProjection`s that use a certain selector?

scothis replied:
I'm ok with this, we can loosen/expand if/when there is demand

scothis commented on README.md:
````diff
@@ -362,12 +367,107 @@ The `$SERVICE_BINDING_ROOT` environment variable **MUST NOT** be reset if it is
+### Custom Projection Definition
+
+```yaml
+apiVersion: projection.internal.service.binding/v1alpha2
````
```suggestion
apiVersion: internal.service.binding/v1alpha2
```

scothis commented on README.md:
```diff
@@ -362,12 +367,107 @@ The `$SERVICE_BINDING_ROOT` environment variable **MUST NOT** be reset if it is
 
+## Custom Projection
+
+There are scenarios where the Reconciler that processes the bindings of a Provisioned Service (hereinafter referred to as `Reconciler A`) is different than the Reconciler that will project those bindings into the Application (hereinafter referred to as `Reconciler B`). To transfer the projection responsibility from Reconciler A to Reconciler B the `ServiceBinding` CR author **MUST** set the `proj ...
```
Clarify that w're not talking about the reconciler of a Projected Service, but of a specific ServiceBinding

```suggestion
There are scenarios where the Reconciler that processes a `ServiceBinding` (hereinafter referred to as `Reconciler A`) is different than the Reconciler that will project the binding into the Application (hereinafter referred to as `Reconciler B`). To transfer the projection responsibility from Reconciler A to Reconciler B the `ServiceBinding` CR author **MUST** set the `projection.service.binding/type` annotation to `Custom`.  
```

arthurdm replied:
good clarification, thanks

scothis commented on README.md:
```diff
@@ -362,12 +367,107 @@ The `$SERVICE_BINDING_ROOT` environment variable **MUST NOT** be reset if it is
+
+There are scenarios where the Reconciler that processes the bindings of a Provisioned Service (hereinafter referred to as `Reconciler A`) is different than the Reconciler that will project those bindings into the Application (hereinafter referred to as `Reconciler B`). To transfer the projection responsibility from Reconciler A to Reconciler B the `ServiceBinding` CR author **MUST** set the `proj ...
+
+Reconciler A reacts to this annotation by creating a new `ServiceBindingProjection` CR which includes the necessary information for Reconciler B, who is watching for and takes ownership of new `ServiceBindingProjection` resources (see the [Custom Projection Definition](#custom-projection-definition) section below), to carry out the projection.  Reconciler A is responsible for updating `ServiceBin ...
```
"ownership" is overloaded and has a different meaning in this context. See https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents

```suggestion
Reconciler A reacts to this annotation by creating a new `ServiceBindingProjection` CR which includes the necessary information for Reconciler B, who is watching and takes responsibility for the new `ServiceBindingProjection` resource (see the [Custom Projection Definition](#custom-projection-definition) section below), to carry out the projection.  Reconciler A is responsible for updating `ServiceBindingProjection`'s spec upon corresponding changes to `ServiceBinding`.
```

arthurdm replied:
thx

arthurdm commented:
thanks for the helpful reviews / commits @scothis!  I believe all of the comments were addressed.

sbose78 commented:
Hi, greetings folks.

Apart from runtime component operator, is any other operator planning to write a reconciler ( or thinking of a similar usecase ) for `ServiceBindingProjection` at the moment ?

scothis commented:
Part of the intent behind custom projection is that we don't need to know. Anybody can hook in at any point in the future without our knowledge.
